src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +621,205 @@ public static string GetTempPath(string workingDir)
+            try
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            }
+            catch (FileNotFoundException)
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+
+            var sw = Stopwatch.StartNew();
+            do
+            {
+                try
+                {
+                    Stream.Lock(0, 0);
+                    IsLocked = true;
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock ()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
+        }
+    }
+
+    internal sealed class ServerNamedMutex : IServerMutex
+    {
+        private readonly Mutex Mutex;
+
+        public bool IsDisposed { get; private set; }
+        public bool IsLocked { get; private set; }
+
+        public ServerNamedMutex (string mutexName, bool createNewLocked, out bool createdNew)
+        {
+            if (createNewLocked) {
+                Mutex = new Mutex(
+                    initiallyOwned: true,
+                    name: mutexName,
+                    createdNew: out createdNew
+                );
+                if (createdNew)
+                    IsLocked = true;
+            } else {
+                Mutex.TryOpenExisting(mutexName, out Mutex);
+                createdNew = false;
+            }
+
+            IsDisposed = (Mutex == null);
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsDisposed)
+                return false;
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+            return IsLocked = Mutex.WaitOne(timeoutMs);
+        }
+
+        public void Unlock ()
+        {
+            if (IsDisposed)
```

Comment:
Should throw `InvalidOperationException` when already disposed.

Reply:
Not ObjectDisposedException?

Reply:
Yeah that's a better choice. Here and everywhere else you check for `IsDisposed`

---
In reply to: [253142709](https://github.com/dotnet/roslyn/pull/31497#discussion_r253142709) [](ancestors = 253142709)

Reply:
Nevermind, it seems like ObjectDisposedException isn't available here, at least according to VS intellisense. I do recall it's in an odd assembly/namespace...

Reply:
That should be there already. I would add it , compile and iterate if that fails. Intellisense is just an approximation of compilation and multi-targeting can wreak havoc sometimes.